tasks/AISTask.cpp

```diff
@@ -36,73 +36,153 @@ bool AISTask::configureHook()
+{
     if (sentence.id() != nmea::sentence_id::VDM) {
         return false;
     }
 
     unique_ptr<marnav::ais::message> msg;
     try {
-        msg = mAIS->processSentence(sentence);
+        msg = m_AIS->processSentence(sentence);
         if (!msg) {
             return true;
         }
     }
     catch (MarnavParsingError const& e) {
         LOG_ERROR_S << "error reported by marnav while creating an AIS message: "
-                    << e.what()
-                    << std::endl;
-        mAISStats.invalid_messages++;
+                    << e.what() << std::endl;
+        m_AIS_stats.invalid_messages++;
         return true;
     }
 
-    mAISStats.received_messages++;
+    m_AIS_stats.received_messages++;
     switch (msg->type()) {
         case ais::message_id::position_report_class_a: {
             auto msg01 = ais::message_cast<ais::message_01>(msg);
-            _positions.write(AIS::getPosition(*msg01));
+            auto position = AIS::getPosition(*msg01);
+            processPositionReport(position, position.mmsi);
             break;
         }
         case ais::message_id::position_report_class_a_assigned_schedule: {
             auto msg02 = ais::message_cast<ais::message_02>(msg);
-            _positions.write(AIS::getPosition(*msg02));
+            auto position = AIS::getPosition(*msg02);
+            processPositionReport(position, position.mmsi);
             break;
         }
         case ais::message_id::position_report_class_a_response_to_interrogation: {
             auto msg03 = ais::message_cast<ais::message_03>(msg);
-            _positions.write(AIS::getPosition(*msg03));
+            auto position = AIS::getPosition(*msg03);
+            processPositionReport(position, position.mmsi);
             break;
         }
         case ais::message_id::static_and_voyage_related_data: {
             auto msg05 = ais::message_cast<ais::message_05>(msg);
-            _vessels_information.write(AIS::getVesselInformation(*msg05));
-            _voyages_information.write(AIS::getVoyageInformation(*msg05));
+            auto vesselInformation = AIS::getVesselInformation(*msg05);
+            auto voyageInformation = AIS::getVoyageInformation(*msg05);
+            updateKnownVessels(vesselInformation);
+            _vessels_information.write(vesselInformation);
+            _voyages_information.write(voyageInformation);
             break;
         }
         default:
-            mAISStats.ignored_messages++;
+            m_AIS_stats.ignored_messages++;
             break;
     }
     return true;
 }
+void AISTask::processPositionReport(ais_base::Position& position, int mmsi)
+{
+    if (!m_use_correction) {
+        _positions.write(position);
+        return;
+    }
+
+    auto vessel = getCorrespondingVesselInfo(mmsi);
+    if (vessel.has_value()) {
+        Eigen::Quaterniond vessel2world_ori(
+            Eigen::AngleAxisd(position.course_over_ground.getRad(),
```

Comment:
The better information is `yaw` for the orientation the vessel in the world. `course_over_ground` should be used only if above a certain value (20cm/s is a good rule of thumb) and if `yaw` is not available.

I think you should add a field to the output struct to indicate the status of the correction

`POSITION_CENTERED_USING_HEADING` / `POSITION_CENTERED_USING_COURSE` / `POSITION_RAW`